pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1506 @@ impl<C: CommandExt> CloseFileDescriptors for C {
+                // shutdown is handled outside of this task
+                match tokio::time::timeout(work_timeout, work_rx.recv_async()).await {
+                    Ok(m) => m.ok(),
+                    Err(_timeout) => {
+                        // we are in a good place to exit, so do it
+                        if feedback.should_downsize() {
+                            // returning None here will exit from the task, and so we can teardown
+                            // this process
+                            None
+                        } else {
+                            continue;
+                        }
+                    }
+                }
+            } else {
+                next
+            };
+
+            match next {
+                Some(t) => {
+                    feedback.dec_waiting_work();
+                    (t, reservation)
+                }
+                None => {
+                    // the Handle or the request queue sender have been dropped; return Ok(()) to keep
+                    // processing any of already pipelined requests.
+                    //
+                    // the drop however does never seem to happen, but most likely the task is
+                    // descheduled because of a shutdown_watcher.
+                    return Ok(());
+                }
             }
+        };
 
-            // If we have some messages in stderr, forward them to the log.
-            let err_revents = pollfds[1].revents().unwrap();
-            if err_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
-                let mut errbuf: [u8; 16384] = [0; 16384];
-                let n = self.stderr.read(&mut errbuf)?;
+        let timeout_at = tokio::time::Instant::now() + request.timeout;
 
-                // The message might not be split correctly into lines here. But this is
-                // good enough, the important thing is to get the message to the log.
-                if n > 0 {
-                    error!(
-                        "wal-redo-postgres: {}",
-                        String::from_utf8_lossy(&errbuf[0..n])
-                    );
+        let records = request.records_range.sub_slice(&request.records);
+
+        if have_vectored_stdin {
+            build_vectored_messages(
+                request.target,
+                request.base_img,
+                records,
+                &mut scratch,
+                &mut buffers,
+            );
+        } else {
+            build_messages(
+                request.target,
+                request.base_img,
+                records,
+                &mut scratch,
+                &mut buffers,
+            );
+        }
+
+        let write_res = async {
+            while buffers.has_remaining() {
+                poll_fn(|cx| poll_write_buf(Pin::new(&mut stdin), cx, &mut buffers))
+                    .await
+                    .map_err(anyhow::Error::new)?;
+            }
```

Comment:
This loop is just distracting.
Why is there no `write_full` method on stdin?
I guess probably because of some `async` Rust limitation.
Can we define an `util method ourselves, then, to declutter this a little?